api/v1beta1/jujucluster_types.go

```diff
@@ -36,7 +40,141 @@ type APIEndpoint struct {
+	// regions, may be overridden by a region.
+	IdentityEndpoint string `json:"identityEndpoint,omitempty"`
+
+	// StorageEndpoint is the default storage endpoint for the cloud
+	// regions, may be overridden by a region.
+	StorageEndpoint string `json:"storageEndpoint,omitempty"`
+
+	// Regions are the regions available in the cloud.
+	//
+	// Regions is a slice, and not a map, because order is important.
+	// The first region in the slice is the default region for the
+	// cloud.
+	Regions []Region `json:"regions,omitempty"`
+
+	// Config contains optional cloud-specific configuration to use
+	// when bootstrapping Juju in this cloud. The cloud configuration
+	// will be combined with Juju-generated, and user-supplied values;
+	// user-supplied values taking precedence.
+	// This should be a map[string]interface{}, but kube-builder does not allow interface{} types
+	// so we are forced to use untyped JSON here
+	Config *apiextensionsv1.JSON `json:"config,omitempty"`
+
+	// RegionConfig contains optional region specific configuration.
+	// Like Config above, this will be combined with Juju-generated and user
+	// supplied values; with user supplied values taking precedence.
+	// This should be a map[string]map[string]interface{}, but kube-builder does not allow interface{} types
+	// so we are forced to use untyped JSON here
+	RegionConfig *apiextensionsv1.JSON `json:"regionConfig,omitempty"`
+
+	// CACertificates contains an optional list of Certificate
+	// Authority certificates to be used to validate certificates
+	// of cloud infrastructure components
+	// The contents are Base64 encoded x.509 certs.
+	CACertificates []string `json:"CACertificates,omitempty"`
+
+	// SkipTLSVerify is true if the client should be asked not to
+	// validate certificates. It is not recommended for production clouds.
+	// It is secure (false) by default.
+	SkipTLSVerify bool `json:"skipTLSVerify,omitempty"`
+
+	// IsControllerCloud is true when this is the cloud used by the controller.
+	IsControllerCloud bool `json:"isControllerCloud,omitempty"`
+}
+
+type Credential struct {
+	// CredentialSecretName is used to tell the controller the name of the secret containing the cloud credentials
+	CredentialSecretName string `json:"credentialSecretName"`
+
+	// CredentialSecretNamespace is used to tell the controller the namespace of the secret containing the cloud credentials
+	CredentialSecretNamespace string `json:"credentialSecretNamespace"`
+}
+
+func deepCopy(dst interface{}, src interface{}) {
+	if dst == nil {
+		panic("dst cannot be nil")
+	}
+	if src == nil {
+		panic("src cannot be nil")
+	}
+	bytes, err := json.Marshal(src)
+	if err != nil {
+		panic("Unable to marshal src")
+	}
+	err = json.Unmarshal(bytes, dst)
+	if err != nil {
+		panic("Unable to unmarshal into dst")
+	}
+}
+
+// We are only able to do this since constraints.Value has json-tagged fields
+type ConstraintValue constraints.Value
```

Comment:
unlike the cloud struct, constraints.Value does not use empty interfaces and DOES have JSON tagged fields, so we can use it here without duplicating its fields